### ISSUE TITLE
Spring Service Registry native-image failure due to service loader handling in jersey

### ISSUE THREAD

**Issue by aclement** [oracle/graal#2652]
**Describe the issue**
I believe native-image is trying to do some eager analysis of service loader usage in my application. Unfortunately jersey seems to bend the spec a little? (I'm not a service loader expert). This causes the native-image handling to exit via exception. My sample app fails like this:

```
Caused by: java.lang.IllegalArgumentException: Class com.sun.jersey.core.impl.provider.entity.DocumentProvider cannot be instantiated reflectively . It does not have a nullary constructor.
        at org.graalvm.nativeimage.hosted.RuntimeReflection.registerForReflectiveInstantiation(RuntimeReflection.java:127)
        at com.oracle.svm.hosted.ServiceLoaderFeature.handleType(ServiceLoaderFeature.java:222)
        at com.oracle.svm.hosted.ServiceLoaderFeature.duringAnalysis(ServiceLoaderFeature.java:112)
        at com.oracle.svm.hosted.NativeImageGenerator.lambda$runPointsToAnalysis$8(NativeImageGenerator.java:720)
        at com.oracle.svm.hosted.FeatureHandler.forEachFeature(FeatureHandler.java:70)
        at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:720)
        at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:538)
        at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:451)
        at java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1386)
        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
```

And indeed in the jersey-core (1.19.1) jar there is a file `META-INF/services/javax.ws.rs.ext.MessageBodyReader` which contains the entry `com.sun.jersey.core.impl.provider.entity.DocumentProvider` which has no default constructor:
```
$ javap -private com.sun.jersey.core.impl.provider.entity.DocumentProvider | grep DocumentProvider

Compiled from "DocumentProvider.java"
public final class com.sun.jersey.core.impl.provider.entity.DocumentProvider extends com.sun.jersey.core.provider.AbstractMessageReaderWriterProvider<org.w3c.dom.Document> {
  public com.sun.jersey.core.impl.provider.entity.DocumentProvider(com.sun.jersey.spi.inject.Injectable<javax.xml.parsers.DocumentBuilderFactory>, com.sun.jersey.spi.inject.Injectable<javax.xml.transform.TransformerFactory>);
```
It only has that custom constructor. I think the jersey infra is applying dependency injection to find those params then calling the custom constructor - and it won't load through normal service loader means.  I can't seem to work around this?

**Steps to reproduce the issue**
Please include both build steps as well as run steps
- git clone https://github.com/aclement/graal-eurekaserver.git
- There are two samples here, at the root is a full blown spring sample that can be built using compile.sh and will show this error. *However*, in the jo subfolder is a much smaller jersey-only sample simply showing the problem:
- `cd jo`
- `mvn compile`
- `mvn exec:java -Dexec.mainClass=Run`
This will display what happens if you just try to use regular service loading on this services file for MessageBodyReader. It should show: `Attempted to load 27 entries: 10 loaded OK, 16 failed missing nullary ctor` (beyond DocumentProvider are numerous other ones that have the same problem)
- `native-image -H:+ReportExceptionStackTraces --no-fallback -cp jsr311-api-1.1.1.jar:jersey-core-1.19.1.jar:jersey-client-1.19.1.jar:target/classes Run` 
Fails with:
```
Caused by: java.lang.IllegalArgumentException: Class com.sun.jersey.core.impl.provider.entity.DocumentProvider cannot be instantiated reflectively . It does not have a nullary constructor.
	at org.graalvm.nativeimage.hosted.RuntimeReflection.registerForReflectiveInstantiation(RuntimeReflection.java:127)
	at com.oracle.svm.hosted.ServiceLoaderFeature.handleType(ServiceLoaderFeature.java:222)
	at com.oracle.svm.hosted.ServiceLoaderFeature.duringAnalysis(ServiceLoaderFeature.java:112)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$runPointsToAnalysis$8(NativeImageGenerator.java:720)
	at com.oracle.svm.hosted.FeatureHandler.forEachFeature(FeatureHandler.java:70)
```

**Describe GraalVM and your environment:**
 - Tried with Graal 20.1 (Java8) and 20.2 dev (20200627) (Java11)
 - macOS Catalina

**More details**
Consider adding the `--native-image-info` and  `--verbose` flags when building your native image and paste output below.

Add any other information about the problem here. Especially important are stack traces or log output. Feel free to link to gists or to screenshots if necessary.
This log is from 20.2 dev Java11 (20200627):
```
Build on Server(pid: 8648, port: 55147)
[run:8648]    classlist:     218.96 ms,  1.67 GB
[run:8648]        (cap):   1,125.84 ms,  1.67 GB
[run:8648]        setup:   1,385.49 ms,  1.67 GB
[run:8648]     analysis:   3,525.15 ms,  1.67 GB
Fatal error:java.lang.IllegalArgumentException: java.lang.IllegalArgumentException: Class com.sun.jersey.core.impl.provider.entity.DocumentProvider cannot be instantiated reflectively . It does not have a nullary constructor.
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:600)
	at java.base/java.util.concurrent.ForkJoinTask.get(ForkJoinTask.java:1006)
	at com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:465)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:346)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:505)
	at com.oracle.svm.hosted.server.NativeImageBuildServer.executeCompilation(NativeImageBuildServer.java:406)
	at com.oracle.svm.hosted.server.NativeImageBuildServer.lambda$processCommand$8(NativeImageBuildServer.java:343)
	at com.oracle.svm.hosted.server.NativeImageBuildServer.withJVMContext(NativeImageBuildServer.java:424)
	at com.oracle.svm.hosted.server.NativeImageBuildServer.processCommand(NativeImageBuildServer.java:340)
	at com.oracle.svm.hosted.server.NativeImageBuildServer.processRequest(NativeImageBuildServer.java:284)
	at com.oracle.svm.hosted.server.NativeImageBuildServer.lambda$serve$7(NativeImageBuildServer.java:244)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.IllegalArgumentException: Class com.sun.jersey.core.impl.provider.entity.DocumentProvider cannot be instantiated reflectively . It does not have a nullary constructor.
	at org.graalvm.sdk/org.graalvm.nativeimage.hosted.RuntimeReflection.registerForReflectiveInstantiation(RuntimeReflection.java:127)
	at com.oracle.svm.hosted.ServiceLoaderFeature.handleType(ServiceLoaderFeature.java:222)
	at com.oracle.svm.hosted.ServiceLoaderFeature.duringAnalysis(ServiceLoaderFeature.java:112)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$runPointsToAnalysis$8(NativeImageGenerator.java:717)
	at com.oracle.svm.hosted.FeatureHandler.forEachFeature(FeatureHandler.java:70)
	at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:717)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:540)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:453)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1407)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
Error: Image build request failed with exit status 1
```

**Comment by peter-hofer** [comment]
@aclement , I think your diagnosis is correct. I'm not really sure how to work around this other than modifying the bytecode of the class, which is probably what the injection mechanism is doing at some point.

**Comment by aclement** [comment]
I saw some native-image invocation using `-H:-UseServiceLoaderFeature`, I haven't tried it yet but was wondering if that'll turn off what GraalVM is trying to do (and yes, leave the onus on me to add some extra config maybe, but at least it might be workable).

**Comment by peter-hofer** [comment]
Yes, `-H:-UseServiceLoaderFeature` will turn off the code that fails here, but it will also disable all automatic support for `ServiceLoader` implementations on the class path.

**Comment by elvizlai** [comment]
same issue when using eureka with grallvm 20.2

**Comment by cstancu** [comment]
Fixed by f64d70db19b4647722a6dfb14d312e043e635146. Now services without a nullary constructor are ignored by the `ServiceLoaderFeature`. At run time trying to load such services will throw a `ServiceConfigurationError` since the type was not registered for reflective instantiation.
